Broken or Flaky test: GarbageCollectorIT

**Test name(s)**
 - org.apache.accumulo.test.functional.GarbageCollectorIT.testRootUniqueMutationDelete

**Describe the failure observed**

An assertion that there were no delete candidates prior to the test code running failed. The relevant test code and assertion:

```java
    // Ensure that no other candidates exist before starting test.
    Iterator<GcCandidate> cIter = ample.getGcCandidates(datalevel);

    ArrayList<GcCandidate> tempCandidates = new ArrayList<>();
    while (cIter.hasNext()) {
      GcCandidate cTemp = cIter.next();
      log.debug("PreExisting Candidate Found: {}", cTemp);
      tempCandidates.add(cTemp);
    }
    assertTrue(tempCandidates.size() == 0);
```

The failure was:
```
[INFO] Running org.apache.accumulo.test.functional.GarbageCollectorIT
[ERROR] Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 262.5 s <<< FAILURE! -- in org.apache.accumulo.test.functional.GarbageCollectorIT
[ERROR] org.apache.accumulo.test.functional.GarbageCollectorIT.testRootUniqueMutationDelete -- Time elapsed: 15.80 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
	at org.apache.accumulo.test.functional.GarbageCollectorIT.testRootUniqueMutationDelete(GarbageCollectorIT.java:328)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Testing Environment:**

 - Version of this project: 3.1.0-SNAPSHOT
 - First commit known to fail (or current commit): Unknown, happened on 85b7d3be9f7c8942fe2ecd2c4a97e335cb150c8b
 - Other environment details:
```
Apache Maven 3.9.0 (9b58d2bad23a66be161c4664ef21ce219c2c8584)
Maven home: /var/lib/jenkins/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.9.0
Java version: 17.0.8, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-17-openjdk-17.0.8.0.7-1.fc38.x86_64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.5.6-200.fc38.x86_64", arch: "amd64", family: "unix"
```

**What have you tried already?**

The IT got re-run, and passed on a subsequent attempt.

**Additional context**

This code is similar in 2.1, so any failure here is probably applicable to 2.1 as well.
The `assertTrue(size == 0)` should also be changed to `assertEquals(0, size)`

--- a/test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorIT.java
+++ b/test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorIT.java
@@ -34,7 +34,6 @@ import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map.Entry;
-import java.util.Objects;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -309,7 +308,6 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
     killMacGc();
     TableId tableId = DataLevel.METADATA.tableId();
     log.info("Root GcCandidate Deletion test");
-    // Behavior for 2.1. INUSE candidates deletion support will be added in 3.x
     log.info("GcCandidates will be added but not removed from Zookeeper");
 
     Ample ample = cluster.getServerContext().getAmple();
@@ -321,10 +319,20 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
     ArrayList<GcCandidate> tempCandidates = new ArrayList<>();
     while (cIter.hasNext()) {
       GcCandidate cTemp = cIter.next();
-      log.debug("PreExisting Candidate Found: {}", cTemp);
       tempCandidates.add(cTemp);
     }
-    assertTrue(tempCandidates.size() == 0);
+    if (tempCandidates.size() != 0) {
+      ample.deleteGcCandidates(datalevel, tempCandidates, Ample.GcCandidateType.VALID);
+      tempCandidates.clear();
+
+      cIter = ample.getGcCandidates(datalevel);
+      while (cIter.hasNext()) {
+        GcCandidate cTemp = cIter.next();
+        log.debug("PreExisting Candidate Found: {}", cTemp);
+        tempCandidates.add(cTemp);
+      }
+      assertEquals(0, tempCandidates.size());
+    }
 
     // Create multiple candidate entries
     List<GcCandidate> candidates =
@@ -347,11 +355,11 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
       counter++;
     }
     // Ensure Zookeeper collapsed the entries and did not support duplicates.
-    assertTrue(counter == 2);
+    assertEquals(2, counter);
 
     cIter = ample.getGcCandidates(datalevel);
     while (cIter.hasNext()) {
-      // This should be a noop call. Root inUse candidate deletions are not supported in 2.1.x
+      // This should be a noop call. Root inUse candidate deletions are not supported.
       ample.deleteGcCandidates(datalevel, List.of(cIter.next()), Ample.GcCandidateType.INUSE);
     }
 
@@ -365,13 +373,13 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
       for (GcCandidate cand : candidates) {
         if (gcC.getPath().equals(cand.getPath())) {
           // Candidate uid's will never match as they are randomly generated in 2.1.x
-          assertTrue(!Objects.equals(gcC.getUid(), cand.getUid()));
+          assertNotEquals(gcC.getUid(), cand.getUid());
           counter--;
         }
       }
     }
     // Ensure that we haven't seen more candidates than we expected.
-    assertTrue(counter == 0);
+    assertEquals(0, counter);
 
     // Delete the candidates as VALID GcCandidates
     cIter = ample.getGcCandidates(datalevel);
@@ -389,7 +397,7 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
         counter++;
       }
     }
-    assertEquals(counter, 0);
+    assertEquals(0, counter);
   }
 
   @Test
@@ -471,7 +479,7 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
       log.debug("PreExisting Candidate Found: {}", cTemp);
       candidates.add(cTemp);
     }
-    assertTrue(candidates.size() == 0);
+    assertEquals(0, candidates.size());
 
     // Create multiple candidate entries
     List<StoredTabletFile> stfs = Stream
@@ -489,7 +497,7 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
       log.debug("Candidate Found: {}", cTemp);
       candidates.add(cTemp);
     }
-    assertTrue(candidates.size() == 2);
+    assertEquals(2, candidates.size());
 
     GcCandidate deleteCandidate = candidates.get(0);
     assertNotNull(deleteCandidate);
@@ -506,12 +514,12 @@ public class GarbageCollectorIT extends ConfigurableMacBase {
       GcCandidate gcC = candidate.next();
       log.debug("Candidate Found: {}", gcC);
       if (gcC.getPath().equals(deleteCandidate.getPath())) {
-        assertTrue(!Objects.equals(gcC.getUid(), deleteCandidate.getUid()));
+        assertNotEquals(gcC.getUid(), deleteCandidate.getUid());
         foundNewCandidate = true;
       }
       counter++;
     }
-    assertTrue(counter == 2);
+    assertEquals(2, counter);
     assertTrue(foundNewCandidate);
   }
 }

